nodes/filter/single.go

```diff
@@ -1,27 +1,345 @@
+	v, vNull := inst.ReadPinAsFloat(node.In)
+	m, mNull := inst.ReadPinAsFloat(node.In1)
+
+	if vNull || mNull {
+		inst.WritePinNull(node.Out)
+	} else {
+		if toFixed(v, 2) == toFixed(m, 2) {
+			inst.WritePinFloat(node.Out, v)
+			inst.lastValue = v
+		} else {
+			inst.WritePinFloat(node.Out, inst.lastValue)
+		}
+	}
+}
+
+// onlyEqualString, output the last known value if two inputs do not equal to each other. Out put nil if any of the inputs are nil.
+// lastValue is defaulted to be "defaultString"
+func (inst *OnlyEqualString) Process() {
+	v, vNull := inst.ReadPinAsString(node.In)
+	m, mNull := inst.ReadPinAsString(node.In1)
+
+	if vNull || mNull {
+		inst.WritePinNull(node.Out)
+	} else {
+		if v == m {
+			inst.WritePin(node.Out, v)
+			inst.lastValue = v
+		} else {
+			inst.WritePin(node.Out, inst.lastValue)
+		}
+	}
+}
+
+// onlyBetween, Output the last known input float value if the input is not in range. Out put nil if any of the inputs are nil.
+// lastValue is defaulted to be 0
+func (inst *OnlyBetween) Process() {
+	v, vNull := inst.ReadPinAsFloat(node.In)
+	min, minNull := inst.ReadPinAsFloat(node.InMin)
+	max, maxNull := inst.ReadPinAsFloat(node.InMax)
+
+	if vNull || minNull || maxNull {
+		inst.WritePinNull(node.Out)
+	} else {
+		if toFixed(min, 2) <= toFixed(v, 2) && toFixed(v, 2) <= toFixed(max, 2) {
+			inst.WritePinFloat(node.Out, v)
+			inst.lastValue = v
+		} else {
+			inst.WritePinFloat(node.Out, inst.lastValue)
+		}
+	}
+}
+
+// onlyGreater, Output the last known input float value if the input is less than 'OutMax'. Out put nil if any of the inputs are nil.
+// lastValue is defaulted to be 0
+func (inst *OnlyGreater) Process() {
+	v, vNull := inst.ReadPinAsFloat(node.In)
+	min, minNull := inst.ReadPinAsFloat(node.InMin)
+
+	if vNull || minNull {
+		inst.WritePinNull(node.Out)
+	} else {
+		if toFixed(v, 2) > toFixed(min, 2) {
+			inst.WritePinFloat(node.Out, v)
+			inst.lastValue = v
+		} else {
+			inst.WritePinFloat(node.Out, inst.lastValue)
+		}
+	}
+}
+
+// onlyLower, Output the last known input float value if the input is greater than 'OutMin'. Out put nil if any of the inputs are nil.
+// lastValue is defaulted to be 0
+func (inst *OnlyLower) Process() {
+	v, vNull := inst.ReadPinAsFloat(node.In)
+	max, maxNull := inst.ReadPinAsFloat(node.InMax)
+
+	if vNull || maxNull {
+		inst.WritePinNull(node.Out)
+	} else {
+		if toFixed(v, 2) < toFixed(max, 2) {
+			inst.WritePinFloat(node.Out, v)
+			inst.lastValue = v
+		} else {
+			inst.WritePinFloat(node.Out, inst.lastValue)
+		}
+	}
+}
+
+// preventDuplicates, Output the last known input float value if the subsequent input value is the same. Out put nil if the input is nil.
+// lastValue is defaulted to be 0
+func (inst *PreventDuplicates) Process() {
+	v, vNull := inst.ReadPinAsFloat(node.In)
+
+	if vNull {
+		inst.WritePinNull(node.Out)
+	} else {
+		if toFixed(v, 2) == inst.lastValue {
+			inst.WritePinFloat(node.Out, inst.lastValue)
```

Comment:
The idea with this node is that we don't want to send duplicate messages.  So if `v==inst.lastValue` then we shouldn't do anything.